toolchain/check/eval.cpp

```diff
@@ -17,6 +17,104 @@
 
 namespace Carbon::Check {
 
+namespace {
+// Information about an eval block of a specific that we are currently building.
+struct SpecificEvalInfo {
+  // The region within the specific whose eval block we are building.
+  SemIR::GenericInstIndex::Region region;
+  // The work-in-progress contents of the eval block.
+  llvm::ArrayRef<SemIR::InstId> values;
+};
+
+// Information about the context within which we are performing evaluation.
+struct EvalContext {
+  auto GetInContext(SemIR::ConstantId const_id) -> SemIR::ConstantId {
+    if (!const_id.is_symbolic()) {
+      return const_id;
+    }
+
+    // While resolving a specific, map from previous instructions in the eval
+    // block into their evaluated values. These values won't be present on the
+    // instance itself yet, so `GetConstantInInstance` won't be able to find
+    // them.
+    if (specific_eval_info) {
+      const auto& symbolic_info =
+          context.constant_values().GetSymbolicConstant(const_id);
+      if (symbolic_info.index.is_valid() &&
+          symbolic_info.generic_id ==
+              context.generic_instances().Get(specific_id).generic_id &&
+          symbolic_info.index.region() == specific_eval_info->region) {
+        auto inst_id = specific_eval_info->values[symbolic_info.index.index()];
+        CARBON_CHECK(inst_id.is_valid())
+            << "Forward reference in eval block: index "
+            << symbolic_info.index.index() << " referenced before evaluation";
+        return context.constant_values().Get(inst_id);
+      }
+    }
+
+    // Map from a specific constant value to the canonical value.
+    return GetConstantInInstance(context.sem_ir(), specific_id, const_id);
+  }
+
+  // Gets the constant value of the specified instruction in this context.
+  auto GetConstantValue(SemIR::InstId inst_id) -> SemIR::ConstantId {
+    return GetInContext(context.constant_values().Get(inst_id));
+  }
+
+  // Gets the constant value of the specified type in this context.
+  auto GetConstantValue(SemIR::TypeId type_id) -> SemIR::ConstantId {
+    return GetInContext(context.types().GetConstantId(type_id));
+  }
+
+  // Gets the instruction describing the constant value of the specified type in
+  // this context.
+  template <typename IdT>
+  auto GetConstantValueAsInst(IdT id) -> SemIR::Inst {
+    return insts().Get(
+        context.constant_values().GetInstId(GetConstantValue(id)));
+  }
```

Comment:
This call confuses me a little: should it be taking a TypeId as parameter? It looks like the templated version isn't used.

(or will more callers be added? for one caller, maybe this function should be inlined)

Reply:
I've ungeneralized it. I'd prefer to not inline it; I found that having a name for this operation made it a lot easier to follow what the code was doing, and I am anticipating more callers.